File: src/interfaces/contract.ts
import type { ClientOptions } from "@stellar/stellar-sdk/contract";
import { CarStatus } from "./car-status";

export interface IBaseContractClient {
  readonly options: ClientOptions;
  toXDR(): string;
}

export interface IRentACarContract extends IBaseContractClient {
  __constructor: ({
    admin,
    token,
  }: {
    admin: string;
    token: string;
  }) => Promise<this>;

  get_admin_balance: () => Promise<this>;

  add_car: ({
    owner,
    price_per_day,
    commission,
  }: {
    owner: string;
    price_per_day: number;
    commission: number;
  }) => Promise<this>;

  return_car: ({
    owner,
  }: {
    owner: string;
  }) => Promise<this>;

  get_car_status: ({ owner }: { owner: string }) => Promise<CarStatus>;

  rental: ({
    renter,
    owner,
    total_days_to_rent,
    amount,
  }: {
    renter: string;
    owner: string;
    total_days_to_rent: number;
    amount: number;
  }) => Promise<this>;

  remove_car: ({ owner }: { owner: string }) => Promise<this>;

  payout_owner: ({
    owner,
    amount,
  }: {
    owner: string;
    amount: number;
  }) => Promise<this>;

  payout_admin: ({
    amount,
  }: {
    amount: number;
  }) => Promise<this>;
}

File: contracts/rent-a-car/src/interfaces/contract.rs
use soroban_sdk::{Address, Env};

use crate::storage::types::{car_status::CarStatus, error::Error};

pub trait RentACarContractTrait {
    fn __constructor(env: &Env, admin: Address, token: Address) -> Result<(), Error>;
    fn add_car(env: &Env, owner: Address, price_per_day: i128, commission: i128) -> Result<(), Error>;
    fn get_car_status(env: &Env, owner: Address) -> Result<CarStatus, Error>;
    fn rental(env: &Env, renter: Address, owner: Address, total_days_to_rent: u32, amount: i128) -> Result<(), Error>;
    fn remove_car(env: &Env, owner: Address) -> Result<(), Error>;
    fn payout_owner(env: &Env, owner: Address, amount: i128) -> Result<(), Error>;
    fn return_car(env: &Env, owner: Address) -> Result<(), Error>;
    fn get_admin_balance(env: &Env) -> Result<i128, Error>;
    fn payout_admin(env: &Env, amount: i128) -> Result<(), Error>;
}